app/hotels/src/map/MapView.js

```diff
@@ -0,0 +1,133 @@
+// @flow
+
+import * as React from 'react';
+import { StyleSheet, View } from 'react-native';
+import { createFragmentContainer, graphql } from 'react-relay';
+import idx from 'idx';
+
+import Map from './Map';
+import HotelSwipeList from './HotelSwipeList';
+import type { MapView as MapViewData } from './__generated__/MapView.graphql';
+
+type Props_ = {|
+  data: MapViewData,
+|};
+
+type Props = Props & $Shape<Props_>;
+
+type State = {|
+  selectedHotelIndex: number,
+|};
+
+const styles = StyleSheet.create({
+  container: {
+    ...StyleSheet.absoluteFillObject,
+    justifyContent: 'flex-end',
+    alignItems: 'center',
+  },
+});
+
+class MapView extends React.Component<Props, State> {
+  constructor(props: Props) {
+    super(props);
+
+    this.state = {
+      selectedHotelIndex: 0,
+    };
+  }
+
+  componentWillReceiveProps = (nextProps: Props) => {
+    if (this.props.data !== nextProps.data) {
+      // TODO keep correct index if node with the same id is still present
+      this.setState({ selectedHotelIndex: 0 });
+    }
+  };
+
+  getHotels = () => {
+    return idx(this.props, _ => _.data.edges) || [];
+  };
+
+  selectMarker = (selectedHotelIndex: number) => {
+    if (this.state.selectedHotelIndex === selectedHotelIndex) {
+      return;
+    }
+
+    this.setState({ selectedHotelIndex });
+  };
+
+  onSelectMarker = (hotelId: string) => {
+    const hotels = this.getHotels();
+    const index = hotels.findIndex(
+      hotel => idx(hotel, _ => _.node.id) === hotelId,
+    );
+
+    if (Number.isFinite(index)) {
+      this.selectMarker(index);
+    }
+  };
+
+  onSnapToItem = (index: number) => {
+    this.selectMarker(index);
+  };
+
+  render = () => {
+    const { selectedHotelIndex } = this.state;
+    const hotels = this.getHotels();
+
+    return (
+      <View style={styles.container}>
+        <Map
+          hotels={hotels}
+          selectedIndex={selectedHotelIndex}
+          onSelectMarker={this.onSelectMarker}
+        />
+        <HotelSwipeList
+          items={hotels}
+          selectedIndex={selectedHotelIndex}
+          onSnapToItem={this.onSnapToItem}
+        />
+      </View>
+    );
+  };
+}
+
+graphql`
+  fragment MapViewHotel on HotelAvailability {
+    id
+    price {
+      amount
+      currency
+    }
+    hotel {
+      name
+      mainPhoto {
+        thumbnailUrl
+      }
+      rating {
+        stars
+      }
+      coordinates {
+        lat
+        lng
+      }
+      address {
+        street
+        city
+        zip
+      }
+    }
+  }
+`;
+
+export default createFragmentContainer(
+  MapView,
+  graphql`
+    fragment MapView on HotelAvailabilityConnection {
+      edges {
+        node {
+          ...MapViewHotel @relay(mask: false)
```

Comment:
This is not exactly what I would expect here. Do we really need it?

Reply:
I'm not sure about this as well, I didn't find proper way how to create nested fragments like in AllHotels, because both swipe cards & markers require multiple parts of `HotelAvailability`, but this enabled me to create `MapScreenHotel` type I can still use among components

Reply:
in another words, Both `MapView` & `HotelSwipeList` (and some of their descendants) rely on `MapScreenHotel` fragment => there is no single FragmentContainer component => it's still useful to define it as fragment for flow checking => requires `@relay(mask: false)`

Reply:
... if there is alternative way, I'm for it, but I found this so far the best

Reply:
I am sorry, I still do not understand. It seems like you are not composing fragments correctly. Why it's not possible to define the fragments in subcomponents and just use them here? Every subcomponent has different requirements and they must define their requirements in the fragment. This is the whole point of Relay.

See: https://facebook.github.io/relay/docs/en/fragment-container.html#container-composition
See: https://facebook.github.io/relay/docs/en/thinking-in-relay.html#data-masking

Reply:
ok, might be my misunderstanding, I'm gonna revisit it tonight ;)